File: examples/zig/src/dllmain.zig
const std = @import("std");
const windows = std.os.windows;
const winxtra = @import("winxtra.zig");

const c = @cImport({
    @cInclude("zefir_ui_api.h");
});

fn hashSourceLocation(comptime loc: std.builtin.SourceLocation) comptime_int {
    var hasher = std.hash.Fnv1a_64.init();
    hasher.update(loc.module);
    hasher.update(loc.file);
    hasher.update(loc.fn_name);
    hasher.update(std.mem.asBytes(&loc.line));
    hasher.update(std.mem.asBytes(&loc.column));
    return hasher.final();
}

// Callback for button activation
fn buttonOnActivate() callconv(std.builtin.CallingConvention.winapi) void {
    // Set up message box data
    var messageBoxData = std.mem.zeroInit(c.Zefir_MessageBoxTextData, .{ .m_title = "Button Released".ptr, .m_text = "Button was released".ptr, .m_alignment = .{
        .m_value = c.Zefir_TextBlock_Alignment_Left,
        .m_hasValue = true,
    } });
    // Open message box
    _ = c.Zefir_openMessageBoxText.?(&messageBoxData);
}

// Callback for rendering window content
fn renderContent() callconv(std.builtin.CallingConvention.winapi) void {
    // Render a button
    var buttonData = std.mem.zeroInit(c.Zefir_ButtonData, .{
        .m_uniqueId = comptime hashSourceLocation(@src()),
        .m_label = "Test Button".ptr,
        .m_info = .{
            .m_value = "This is a test button".ptr,
            .m_hasValue = true,
        },
        .m_onActivate = .{
            .m_value = buttonOnActivate,
            .m_hasValue = true,
        },
    });

    // Render the button
    _ = c.Zefir_renderButton.?(&buttonData);
}

fn mainEntry(_: windows.LPVOID) callconv(std.builtin.CallingConvention.winapi) windows.DWORD {
    // Initialize the Zefir API
    // Replace "ToastedMarshmellow.dll" with the name of the zefir UI DLL you want to use
    const success = c.Zefir_initializeAPI_Create("ToastedMarshmellow.dll");
    if (success) {
        var windowData = std.mem.zeroInit(c.Zefir_WindowData, .{
            .m_id = "ZigTest.TestWindow".ptr,
            .m_title = "Zig Window".ptr,
            .m_onRenderContent = .{
                .m_value = renderContent,
                .m_hasValue = true,
            },
        });

        // Add the window
        _ = c.Zefir_addWindow.?(&windowData);
    }
    return std.os.windows.S_OK;
}

pub fn DllMain(
    _: windows.HINSTANCE,
    fdwReason: windows.DWORD,
    _: windows.LPVOID,
) callconv(std.builtin.CallingConvention.winapi) windows.BOOL {
    switch (fdwReason) {
        winxtra.DLL_PROCESS_ATTACH => {
            _ = winxtra.CreateThread(null, 0, mainEntry, null, 0, null);
        },
        winxtra.DLL_PROCESS_DETACH => {},
        winxtra.DLL_THREAD_ATTACH => {},
        winxtra.DLL_THREAD_DETACH => {},
        else => {},
    }

    return windows.TRUE;
}

File: examples/cpp/zefir_cpp_example/dllmain.cpp
// dllmain.cpp : Defines the entry point for the DLL application.
#include "pch.h"
#include "../../../include/zefir_ui_api.h"

#ifdef __cplusplus

static void ZEFIR_CALLBACK showMessageBoxCpp()
{
    zefir::api::popup::text({
        .m_title = "Cpp Popup",
        .m_text = "Hello from C++",
        .m_alignment = {.m_value = Zefir_TextBlock_Alignment_Center, .m_hasValue = true} });
}

static void ZEFIR_CALLBACK renderContentCpp()
{
    zefir::api::cm::text("Hello from C++");
    zefir::api::cm::text("This is a test window");

    zefir::api::cm::button({
        .m_uniqueId = ZEFIR_NEW_UID(),
        .m_label = "Click me",
        .m_onActivate = {.m_value = &showMessageBoxCpp, .m_hasValue = true} });
}

static void ZEFIR_CALLBACK addWindowCpp()
{
    zefir::api::window::add({
        .m_id = "CppTest.TestWindowCpp",
        .m_title = "CPP Window",
        .m_onRenderContent = {.m_value = &renderContentCpp, .m_hasValue = true} });
}

#endif 

static void ZEFIR_CALLBACK showMessageBoxC()
{
    Zefir_PopupTextData data{
        .m_title = "C Popup",
        .m_text = "Hello from C! Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
        .m_customWidth = {.m_value = 450, .m_hasValue = true},
        .m_alignment = {.m_value = Zefir_TextBlock_Alignment_Left, .m_hasValue = true} };
        
    Zefir_openPopupText(&data);
}

static void ZEFIR_CALLBACK renderContentC()
{
    Zefir_renderText("Hello from C!");
    Zefir_renderText("This is a test of the C API.");

    Zefir_ButtonData data{
        .m_uniqueId = ZEFIR_NEW_UID(),
        .m_label = "Click me",
        .m_onActivate = {.m_value = &showMessageBoxC, .m_hasValue = true} };

    Zefir_renderButton(&data);
}

static void ZEFIR_CALLBACK addWindowC()
{
    Zefir_WindowData windowData = {
        .m_id = "CppTest.TestWindowC",
        .m_title = "C Window",
        .m_onRenderContent = {.m_value = &renderContentC, .m_hasValue = true} };

    Zefir_addWindow(&windowData);
}

DWORD WINAPI mainThread(LPVOID lpParam)
{
    if (Zefir_initializeAPI_Create("ToastedMarshmellow.dll"))
    {
#       ifdef __cplusplus
        addWindowCpp();
#       endif
        addWindowC();
    }
    return S_OK;
}

BOOL APIENTRY DllMain(HMODULE hModule,
    DWORD  ul_reason_for_call,
    LPVOID lpReserved
)
{
    switch (ul_reason_for_call)
    {
    case DLL_PROCESS_ATTACH:
        CreateThread(NULL, 0, mainThread, NULL, 0, NULL);
        break;
    case DLL_THREAD_ATTACH:
        break;
    case DLL_THREAD_DETACH:
        break;
    case DLL_PROCESS_DETACH:
        break;
    }
    return TRUE;
}

